### PR TITLE
[typescript][angular] basePath through injected config

### PR DESCRIPTION
**Pull request by akehir**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.1.x`, `4.0.x`. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

@macjohnny : This resolves https://github.com/OpenAPITools/openapi-generator/issues/282 , implementing my swagger-codegen pull request on the fork.

I notice, that there are a few tests missing, notably the ones I've previously added for angular 6. 

CI is here: https://travis-ci.org/akehir/openapi-generator
The original PR here: https://github.com/swagger-api/swagger-codegen/pull/8305

### Original description of the PR
Instead of using the this.basepath, we use this.configuration.basepath - allowing us to inject a different configuration (after loading a configuration file over http during APP_INITIALIZER).
We've been using this setup at a project for a while and it has been working well for us (different api url's in different environments, provided by the webserver on runtime).

This would close the problems discussed here: #3221 (comment) and #7949.

This the distinction to the existing PR #7965 is, that it is not required to generate / handle an observable, but rather just by changing the injected configuration, another basePath would be used. Do you think this would solve your concerns @chrismellard?

Since the rest of the swagger api client is always referring to the this.configuration.{...}, I believe this approach might be more consistent with the general handling of the configuration service.

To keep compatibility with implementations that don't inject a configuration, but just the base path, I set the base path on the internal configuration object to the injected basepath (when they don't exist).

This is the new PR for https://github.com/OpenAPITools/openapi-generator/pull/289

### CODE REVIEW

File: samples/client/petstore/typescript-angular-v6-provided-in-root/pom.xml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                        </goals>
+                        <configuration>
+                            <executable>npm</executable>
+                            <arguments>
+                                <argument>test</argument>
+                                <argument>--</argument>
+                                <argument>--progress=false</argument>
+                                <argument>--no-watch</argument>
+                                <argument>--browsers</argument>
+                                <argument>ChromeHeadless</argument>
```
Comment (macjohnny) [inline samples/client/petstore/typescript-angular-v6-provided-in-root/pom.xml]: glad it worked with ChromeHeadless

**Comment by macjohnny** [comment]
@wing328 any idea where to put a note about the `client/petstore/typescript-angular-v6-provided-in-root/tests/default` sample project that can be used by a developer to test changes, etc.?
Should it be noted in the Readme generated?

**Comment by akehir** [comment]
@macjohnny : The correct place would probably be the Testing section in contributing.md: https://github.com/OpenAPITools/openapi-generator/blob/master/CONTRIBUTING.md#testing

The sample there is listing how it works for ruby, but angular is equivalent.

Run either ```./bin/typescript-angular-petstore-all.sh``` or ```.\bin\windows\typescript-angular-petstore-all.bat``` to generate the samples.

Run either ```mvn verify -Psamples -fae``` (complete CI tasks) or ```mvn integration-test -Psamples -fae``` (to just run the tests).

Alternatively, for just angular language, ```mvn integration-test -Psamples -rf :TypeScriptAngular6PestoreClientTests```.

**Comment by macjohnny** [comment]
@akehir you are right. Would you mind updating your PR and adding it to the contributing.md file?

**Comment by wing328** [comment]
> @akehir you are right. Would you mind updating your PR and adding it to the contributing.md file?

Shall we document that in the wiki page instead?

If we document the test procedure for other generators in the contributing guideline, the page will become too large and likely potential contributors will just skip it.

**Comment by akehir** [comment]
@wing328 @macjohnny : Where's the wiki - here on github? If you point me to the correct location, I'll add the information there.

An alternative is that we create a readme.md / contributing.md within the typescript-angular template folder itself.

**Comment by macjohnny** [comment]
@wing328 the wiki repository https://github.com/OpenAPITools/openapi-generator.wiki.git does not seem to be public.

**Comment by jmini** [comment]
@macjohnny, yes it is. To read it the URL is https://github.com/OpenAPITools/openapi-generator/wiki

To clone it the URL you have provided works:

```
git clone https://github.com/OpenAPITools/openapi-generator.wiki.git
```

(without authentication)

**Comment by macjohnny** [comment]
@jmini how can you file a PR to change the wiki?

**Comment by macjohnny** [comment]
@wing328 could you add this to the 3.0.3 milestone?

**Comment by wing328** [comment]
cc the TS technical committee:

@TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01)

**Comment by wing328** [comment]
I've just restarted the CircleCI job. If it passes, I'll merge this PR into master.

**Pull request merged**